compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -0,0 +1,608 @@
+  for (auto h : handles) b.create<PrintOp>(h);
+}
+
+namespace {
+struct BatchFusionSpec {};
+struct IterativeFusionSpec {};
+
+struct TileOrNumThreadHandle {};
+struct TileOrNumThreadList {};
+}  // namespace
+
+/// Performs the following transformations:
+///   1. Tiles `rootH` to scf.foreach_thread to with `tileSizesOrNumThreads`
+///      according to whether spec is a TileSizesSpec or a NumThreadsSpec.
+///   2. Maps the resulting scf.foreach_thread to threads according to
+///      `threadDimMapping`.
+///   3. Iterates over `opsHToFuse` in order and fuses into the containing op.
+/// Returns a handle to the resulting scf.foreach_thread.
+///
+/// Fusion is controlled by the BatchOrIterativeFusion templatetype:
+///   1. In batch mode, a single fusion command is issued and a topological sort
+///      is automatically computed by the fusion. Since this applies a single
+///      fusion, no interleaved canonicalization/cse/enabling occur and the
+///      resulting fusion may not be as good.
+///   2. In iterative mode, the user is responsible for providing the fusion
+///      order but interleaved canonicalization/cse/enabling occur which may
+///      result in better fusion results.
+///      TODO: Transform dialect op to apply topological sort and avoid the
+///      manual intervention.
+/// If `resultingFusedOpsHandles` is a non-null pointer, the fused operation are
+/// appended in order.
+// TODO: apply forwarding pattern.
+template <typename TilingTransformOp, typename TileOrNumThreadSpec,
+          typename BatchOrIterativeFusion, typename TileOrNumThreadKind>
+static Value tileAndFuseAndDistributeImpl(
+    ImplicitLocOpBuilder &b, Value rootH, ValueRange opsHToFuse,
+    std::conditional_t<std::is_same_v<TileOrNumThreadKind, TileOrNumThreadList>,
+                       ArrayRef<int64_t>, ArrayRef<OpFoldResult>>
+        tileSizesOrNumThreads,
+    ArrayAttr threadDimMapping, SmallVector<Value> *resultingFusedOpsHandles) {
+  auto tileToForeachOp = b.create<TilingTransformOp>(
+      rootH, tileSizesOrNumThreads, TileOrNumThreadSpec(), threadDimMapping);
+  Value foreachThreadH = tileToForeachOp.getForeachThreadOp();
+  if (std::is_same<BatchOrIterativeFusion, BatchFusionSpec>::value) {
+    Value mergedOpsH =
+        b.create<MergeHandlesOp>(opsHToFuse, /*deduplicate=*/true);
+    Value fusedH = b.create<FuseIntoContainingOp>(mergedOpsH, foreachThreadH);
+    (void)fusedH;
+    assert(!resultingFusedOpsHandles && "Handle needs unpacking");
+  } else {
+    for (Value h : opsHToFuse) {
+      Value fusedH = b.create<FuseIntoContainingOp>(h, foreachThreadH);
+      if (resultingFusedOpsHandles) resultingFusedOpsHandles->push_back(fusedH);
+    }
+  }
+  return foreachThreadH;
+}
+
+/// Call tileAndFuseAndDistributeImpl with ArrayRef<int64_t> tilesSizes.
+template <typename TilingTransformOp = TileToForeachThreadOp,
+          typename BatchOrIterativeFusion = BatchFusionSpec>
+static Value buildTFDWithTileSizes(
```

Comment:
Took a while to decode `TFD` as `TileFuseAndDistribute`. Please add comment for that. Also is the templated form necessary.

Reply:
`buildTileFuseDistWithTileSizes`? :)